File: example/xx_three/wip/interact_drag.cpp
module;
#include <infra/Cpp20.h>
#include <xx_three/ex.h>
module two.xxthree;

using namespace two;

EX(xx_interact_drag)
{
#if UI
	UNUSED(dockbar);
	SceneViewer& viewer = ui::scene_viewer(parent);
	Scene& scene = viewer.m_scene;
#else
	static Scene scene = Scene(app.m_gfx);
	static GfxViewer viewer = GfxViewer(window, scene);
#endif

#if UI
	TrackballController& controls = ui::trackball_controller(viewer);
	controls.m_staticMoving = true;
	controls.m_dynamicDampingFactor = 0.3f;
#endif

	static vector<Node3*> objects;

	if(init)
	{
		Camera& camera = viewer.m_camera;
		camera.m_fov = 70.f; camera.m_near = 1.f; camera.m_far = 5000.f;
		camera.m_eye.z = 1000.f;

		//scene.background = new THREE.Color(0xf0f0f0);

		//scene.add(new THREE.AmbientLight(0x505050));

		//var light = new THREE.SpotLight(0xffffff, 1.5);
		//light.position.set(0, 500, 2000);
		//light.angle = c_pi / 9;
		//
		//light.castShadow = true;
		//light.shadow.camera.near = 1000;
		//light.shadow.camera.far = 4000;
		//light.shadow.mapSize.width = 1024;
		//light.shadow.mapSize.height = 1024;
		//
		//scene.add(light);

		Model& geometry = app.m_gfx.shape(Cube(20.f));

		for(size_t i = 0; i < 200; i++)
		{
			vec3 p = vec3(randf() * 1000.f - 500.f, randf() * 600.f - 300.f, randf() * 800.f - 400.f);
			vec3 a = vec3(randf(), randf(), randf()) * c_2pi;
			vec3 s = vec3(randf(), randf(), randf()) * 2.f + 1.f;

			Material& material = gfx::pbr_material(app.m_gfx, "drag" + to_string(i), rgb(randi<uint32_t>()));
			Node3& n = gfx::nodes(scene).add(Node3(p, quat(a), s));
			gfx::items(scene).add(Item(n, geometry, 0U, &material));

			objects.push_back(&n);
		}

		//renderer.shadowMap.enabled = true;
		//renderer.shadowMap.type = THREE.PCFShadowMap;

		//var dragControls = new THREE.DragControls(objects, camera, renderer.domElement);
		//dragControls.addEventListener('dragstart', function() {
		//
		//	controls.enabled = false;
		//
		//});
		//dragControls.addEventListener('dragend', function() {
		//
		//	controls.enabled = true;
		//
		//});

	}
}

File: example/xx_three/wip/interact_drag.js
var viewer = two.ui.scene_viewer(panel);

TrackballController controls = two.ui.trackball_controller(viewer);
controls.staticMoving = true;
controls.dynamicDampingFactor = 0.3;

var scene = viewer.scene;

vector<Node3*> objects;

bool once = false;
if(!once)
{
    once = true;

    var camera = viewer.camera;
    camera.fov = 70.0; camera.near = 1.0; camera.far = 5000.0;
    camera.eye.z = 1000.0;

    //scene.background = new THREE.Color(0xf000);

    //scene.add(new THREE.AmbientLight(0x505050));

    //var light = new THREE.SpotLight(0xffffff, 1.5);
    //light.position.set(0, 500, 2000);
    //light.angle = PI / 9;
    //
    //light.castShadow = true;
    //light.shadow.camera.near = 1000;
    //light.shadow.camera.far = 4000;
    //light.shadow.mapSize.width = 1024;
    //light.shadow.mapSize.height = 1024;
    //
    //scene.add(light);

    var geometry = app.gfx.shape(new two.Cube(20.0));

    for(var i = 0; i < 200; i++)
    {
        var p = new two.vec3(Math.random() * 1000.0 - 500.0, Math.random() * 600.0 - 300.0, Math.random() * 800.0 - 400.0);
        var a = new two.vec3(Math.random(), Math.random(), Math.random()) * 2 * Math.PI;
        var s = new two.vec3(Math.random(), Math.random(), Math.random()) * 2.0 + 1.0;

        var material = two.gfx.pbr_material(app.gfx, 'material' + to_string(i), two.rgb(Math.random() * 0xffffff));
        var n = scene.nodes().add(new two.Node3(p, new two.quat(a), s));
        scene.items().add(new two.Item(n, geometry, 0, material));

        objects.push(n);
    }

    //renderer.shadowMap.enabled = true;
    //renderer.shadowMap.type = THREE.PCFShadowMap;

    //var dragControls = new THREE.DragControls(objects, camera, renderer.domElement);
    //dragControls.addEventListener('dragstart', function() {
    //
    //	controls.enabled = false;
    //
    //});
    //dragControls.addEventListener('dragend', function() {
    //
    //	controls.enabled = true;
    //
    //});

}
